numpy/_core/src/umath/stringdtype_ufuncs.cpp

```diff
@@ -1303,47 +1222,30 @@ replace_resolve_descriptors(struct PyArrayMethodObject_tag *NPY_UNUSED(method),
     PyArray_StringDTypeObject *descr2 = (PyArray_StringDTypeObject *)given_descrs[1];
     PyArray_StringDTypeObject *descr3 = (PyArray_StringDTypeObject *)given_descrs[2];
 
-    // _eq_comparison has a short-circuit pointer comparison fast path, so
-    // no need to do it here
-    int eq_res = (_eq_comparison(descr1->coerce, descr2->coerce,
-                                 descr1->na_object, descr2->na_object) &&
-                  _eq_comparison(descr1->coerce, descr3->coerce,
-                                 descr1->na_object, descr3->na_object));
+    PyArray_StringDTypeObject *common_descr = common_instance(
+            common_instance(descr1, descr2), descr3);
 
-    if (eq_res < 0) {
+    if (common_descr == NULL) {
         return (NPY_CASTING)-1;
     }
 
-    if (eq_res != 1) {
-        PyErr_SetString(PyExc_TypeError,
-                        "String replace is only supported with equal StringDType "
-                        "instances.");
-        return (NPY_CASTING)-1;
-    }
-
-    Py_INCREF(given_descrs[0]);
-    loop_descrs[0] = given_descrs[0];
-    Py_INCREF(given_descrs[1]);
-    loop_descrs[1] = given_descrs[1];
-    Py_INCREF(given_descrs[2]);
-    loop_descrs[2] = given_descrs[2];
-    Py_INCREF(given_descrs[3]);
-    loop_descrs[3] = given_descrs[3];
+    loop_descrs[0] = NPy_NewDescrRef(given_descrs[0]);
+    loop_descrs[1] = NPy_NewDescrRef(given_descrs[1]);
+    loop_descrs[2] = NPy_NewDescrRef(given_descrs[2]);
+    loop_descrs[3] = NPy_NewDescrRef(given_descrs[3]);
 
     PyArray_Descr *out_descr = NULL;
 
     if (given_descrs[4] == NULL) {
         out_descr = (PyArray_Descr *)new_stringdtype_instance(
```

Comment:
Why not use `common_instance` directly? It is a new reference already, no? (and DECREF in other branch)